tools/flow_grid.py

```diff
@@ -0,0 +1,689 @@
+    nodata : int or float (optional)
+             Value indicating no data.
+
+    Attributes (Optional)
+    ---------------------
+    dem : digital elevation grid
+    dir : flow direction grid
+    acc : flow accumulation grid
+    catch : Catchment delineated from 'dir' and a given pour point
+    frac : fractional contributing area grid
+    bbox : The geographical bounding box of the gridded dataset
+           (xmin, ymin, xmax, ymax)
+    shape : The shape of the gridded data (nrows, ncolumns)
+    cellsize : The length/width of each grid cell (assumed to be square).
+    nodata : The value to use for gridcells with no data.
+
+    Methods
+    -------
+    read_input : add a gridded dataset (dem, flowdir, accumulation) 
+                 to flow_grid instance.
+    nearest_cell : Returns the index (column, row) of the cell closest
+                   to a given geographical coordinate (x, y).
+    flowdir : Generate a flow direction grid from a given digital elevation
+              dataset (dem).
+    catchment : Delineate the watershed for a given pour point (x, y)
+                or (column, row).
+    fraction : Generate the fractional contributing area for a coarse
+               scale flow direction grid based on a fine-scale flow
+               direction grid.
+    """
+
+    def __init__(self, conform_data=True, **kwargs):
+        self.conform_data = conform_data
+        self.nodata = {}
+        self.gridlist = []
+        if 'data' in kwargs:
+            self.read_input(**kwargs)
+        else:
+            pass
+
+    def read_input(self, data, data_type='dir', input_type='ascii', band=1,
+                   nodata=0, bbox=None, crs=None, **kwargs):
+        """
+        Reads data into a named attribute of flow_grid
+        (name of attribute determined by 'data_type').
+
+        Parameters
+        ----------
+        data : File name (string) or numpy ndarray
+               If data is from a file, 'input_type' should be set to the
+               appropriate value ('ascii' or 'raster').
+        data_type : 'dem', 'dir', 'acc' or other string
+                     Name of dataset. Will determine the name of the attribute
+                     representing the gridded data. Default values are used
+                     internally by some class methods:
+                         'dem' : digital elevation data
+                         'dir' : flow direction data
+                         'acc' : flow accumulation (upstream area) data
+                    
+        input_type : 'raster', 'ascii' or 'array'
+                     Type of input data.
+        band : int
+               For raster data, the band number to read.
+        nodata : int or float
+                 Value indicating no data.
+        bbox : tuple or list
+               Bounding box, if none provided.
+
+        """
+
+        # read ascii file
+        if input_type == 'ascii':
+            with open(data) as header:
```

Comment:
check.

ascii is now handled by `read_ascii` and raster is now handled by `read_raster`. These functions pass the data and metadata into a generic reader function called `add_data`.